drivers/adc/adc_demo/adc_demo.c

```diff
@@ -0,0 +1,139 @@
+ * from this software without specific prior written permission.
+ * - The use of this software may or may not infringe the patent rights
+ * of one or more patent holders. This license does not release you
+ * from the requirement that you obtain separate licenses from these
+ * patent holders to use this software.
+ * - Use of the software either in source or binary form, must be run
+ * on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include <inttypes.h>
+#include <stdlib.h>
+#include <string.h>
+#include "adc_demo.h"
+#include "error.h"
+#include "util.h"
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/***************************************************************************//**
+ * @brief init function for the adc demo driver
+ *
+ * @param desc - descriptor for the adc
+ *
+ * @param param - initialization param for adc
+ *
+ * @return SUCCESS in case of success, negative error code otherwise.
+*******************************************************************************/
+int32_t adc_demo_init(struct adc_demo_desc **desc,
+		      struct adc_demo_init_param *param)
+{
+	struct adc_demo_desc *adesc;
+	adesc = (struct adc_demo_desc*)calloc(1,sizeof(*adesc));
+
+	if(!adesc)
+		return -ENOMEM;
+
+	*desc = adesc;
+
+	return SUCCESS;
+}
+
+/***************************************************************************//**
+ * @brief free allocated resources
+ *
+ * @param desc - descriptor for the adc
+ *
+ * @return SUCCESS in case of success, FAILURE otherwise.
+*******************************************************************************/
+int32_t adc_demo_remove(struct adc_demo_desc *desc)
+{
+	if(!desc)
+		return FAILURE;
```

Comment:
We now avoid using FAILURE and use error codes from <errno.h>. If you can't find what you want in there, you can create more in our "error.h". In this case it should be `return -EINVAL` (mind the minus -)

You will probably find LOTS of `return FAILURE` occurences throughout older no-OS code but this is for historical reasons... ;)

Reply:
Corrected :)